enhancements/ingress/host-network-configurable-ports.md

```diff
@@ -0,0 +1,298 @@
+
+The status of the Ingress Controller should report if it wasn't able to run the requested number of replicas because of the unavailability of the requested ports on the available nodes.
+
+### User Stories
+
+#### As a cluster administrator, I need to set binding ports for the IngressController so that I can deploy multiple ingress controllers on the same node for the HostNetwork strategy
+
+The cluster administrator can set the ingress controller's `spec.endpointPublishingStrategy.hostNetworkParameters`.
+
+#### As an OpenShift developer, I want to make sure Ingress Controller documentation is updated with this new feature concerning the HostNetwork strategy
+
+#### As an OpenShift developer, automated e2e tests cover the case of deploying 2 ingress controllers in HostNetwork strategy to the same nodes with different sets of ports, so that I can validate the feature
+
+#### As a Cluster Administrator, I want the IngressController status to report problems provisioning the Ingress Controller pods in HostNetwork strategy when there is a port conflict
+
+### API Extensions
+
+This  proposal edits `IngressController.operator.openshift.io/v1`.
+
+### Operational Aspects of API Extensions
+
+N/A.
+
+### Implementation Details
+
+Implementing this needs changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+
+OpenShift Cluster Ingress Operator, creates a deployment for Router with environment variables for port bindings which OpenShift Router already respects:
+`ROUTER_SERVICE_HTTP_PORT`, `ROUTER_SERVICE_HTTPS_PORT`, `STATS_PORT`.
+
+Also the deployments should not have ip:port conflicts for `SNI_PORT` and `NO_SNI_PORT`. There are multiple ways to achieve this:
+
+- The cluster ingress operator configures the router to use a unique loopback address for the sni/nosni frontends. This loopback address is configured automatically, for example by hashing the ingresscontroller’s name.
+  As a result, the sni/nosni port numbers wouldn’t need to change to avoid the conflict as the routers would be using different IP addresses. This can be done with the introduction of two new environment variables for the router:`ROUTER_SERVICE_SNI_IP`, `ROUTER_SERVICE_SNI_IP`, and the configuration of the frontend to bind on this IPs.
+  The cluster ingress operator sets these environment variables for the router deployment from `127.0.0.1/8` CIDR based on the ingress controller name hash.
+- The router gets rid of the loopback hop using a Unix domain socket or other solution, which would have other advantages as well ([router's PR](https://github.com/openshift/router/pull/326)).
+
+### Risks and Mitigations
+
+#### Failure Modes
+
+N/A.
+
+#### Support Procedures
+n
+N/A.
+
+#### Port Conflict
+
+Users might define conflicting ports which would cause HAProxy process to fail at startup.
+
+A mitigation to this risk is to implement a validation feature in the reconciliation loop of the Cluster Ingress Operator
```

Comment:
So we're going to validate in the reconcile loop. It would be nicer to prevent the creation of `IngressController` resources with conflicting ports. For instance when a  __NodePort__ service is created with a port which is already allocated then the creation fails with an error message, like so:

> The Service "service02" is invalid: spec.ports[0].nodePort: Invalid value: 30007: provided port is already allocated

Reply:
Yeah, would be good to have a validating webhook for the ingress operator. But for the moment the validation is done in [the reconcile loop](https://github.com/openshift/cluster-ingress-operator/blob/0d5f49acfb118d7a6c553709b730ab1e7b2a6f6c/pkg/operator/controller/ingress/controller.go#L284).